### PR TITLE
LIZA-103: TimeRecognizer: Скрипт обучения

### PR DESCRIPTION
**Pull request by ol00lo**

### CODE REVIEW

File: clock/study.py
Comment (kalininei) [inline clock/study.py:1]: train.py

File: clock/study.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+train_datagen = ImageDataGenerator(
+    rescale=1.0/255.0,
+    width_shift_range=0.2,
+    height_shift_range=0.2,
+    shear_range=0.2,
+    zoom_range=0.2,
+    fill_mode='nearest'
+)
+
+train_generator = train_datagen.flow_from_directory(
```
Comment (kalininei) [inline clock/study.py:22]: А где у вас тут используется imgaug для аугментации?
Мы же говорили, надо написать собственный генератор, а не пользоваться ImageDataGenerator.
И ground-truth надо считывать из файлов с аннотоацией.

File: clock/study.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+base_model = MobileNet(weights='imagenet', include_top=False, input_shape=(IMG_HEIGHT, IMG_WIDTH, 3))
+base_model.trainable = False
+
+model = models.Sequential([
+    base_model,
+    layers.GlobalAveragePooling2D(),
+    layers.Dense(128, activation='relu'),
+    layers.Dropout(0.5), 
+    layers.Dense(train_generator.num_classes, activation='softmax')
```
Comment (kalininei) [inline clock/study.py:37]: У нас не задача классификации, поэтому num_classes не надо.
У нас регрессия. На выходе одно число для одной картинки. 0 -- это 0:00, 1 -- 12:00.

File: clock/study.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+model = models.Sequential([
+    base_model,
+    layers.GlobalAveragePooling2D(),
+    layers.Dense(128, activation='relu'),
+    layers.Dropout(0.5), 
+    layers.Dense(train_generator.num_classes, activation='softmax')  
+])
+
+def custom_loss(y_true, y_pred):
+    y_pred = tf.math.floormod(y_pred, 720)
```
Comment (kalininei) [inline clock/study.py:41]: Тут я не понял, что происходит. Нужна такая функция, которая бы
```
loss(0, 0) = loss(1, 0) -> 0
loss(0.5, 0.8) = loss(0.1, 0.8) -> \approx 0.3
и т.п.
```

File: clock/study.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    diff = tf.abs(y_true - y_pred)
+    loss = tf.reduce_mean(tf.minimum(diff, 720 - diff))
+    return loss
+
+model.compile(optimizer='adam', loss=custom_loss, metrics=['accuracy'])
+
+log_dir = "logs/fit/" + tf.datetime.now().strftime("%Y%m%d-%H%M%S")
+tensorboard_callback = callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)
+
+model_checkpoint = callbacks.ModelCheckpoint(filepath='best_model.keras', save_best_only=True, monitor='val_accuracy')
```
Comment (kalininei) [inline clock/study.py:52]: Пока напишите по лоссу.
И в имени файла записывайте эпоху и значение метрики для удобства

File: clock/study.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+])
+
+def custom_loss(y_true, y_pred):
+    y_pred = tf.math.floormod(y_pred, 720)  
+    y_true = tf.math.floormod(y_true, 720)  
+    diff = tf.abs(y_true - y_pred)
+    loss = tf.reduce_mean(tf.minimum(diff, 720 - diff))
+    return loss
+
+model.compile(optimizer='adam', loss=custom_loss, metrics=['accuracy'])
```
Comment (kalininei) [inline clock/study.py:47]: У нас нет классической метрики `accuracy`. Потому что у нас задача регрессии. 
Пока уберите доп. метрики. Потом подумаем, как можно было бы аналого метрики accuracy для этой задачи придумать.

File: clock/study.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+train_generator = train_datagen.flow_from_directory(
+    DATA_DIR,
+    target_size=(IMG_HEIGHT, IMG_WIDTH),
+    batch_size=BATCH_SIZE,
+    class_mode=None
+)
+
+base_model = MobileNet(weights='imagenet', include_top=False, input_shape=(IMG_HEIGHT, IMG_WIDTH, 3))
+base_model.trainable = False
```
Comment (kalininei) [inline clock/study.py:30]: вы же понимаете да, зачем вы это написали? ))

File: clock/study.py (hunk 1 above)
Comment (ol00lo) [reply clock/study.py]: ✔️

File: clock/study.py
Comment (ol00lo) [reply clock/study.py:1]: ✔️

File: clock/train.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+print(f"144.0 = {custom_loss(true, pred)}")
+true = np.array([710])
+pred = np.array([350.0])
+print(f"a lot = {custom_loss(true, pred)}")
+true = np.array([1.0])
+pred = np.array([2.0])
+print(f"36.0 = {custom_loss(true, pred):.4f}")
+
+model.compile(optimizer='adam', loss=custom_loss)
+log_dir = os.path.join("logs", "fit", f"model_run_{int(time.time())}")
```
Comment (kalininei) [inline clock/train.py:125]: Вам так  удобно будет? Для каждого запуска перезапускть тензорбоард для новой директории? Или удобнее удалять старый лог. Можно прям в скрипте.
Если лог важен для сравнения, его можно руками переименовать.

File: clock/train.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+pred = np.array([719.0])
+print(f"144.0 = {custom_loss(true, pred)}")
+true = np.array([710])
+pred = np.array([350.0])
+print(f"a lot = {custom_loss(true, pred)}")
+true = np.array([1.0])
+pred = np.array([2.0])
+print(f"36.0 = {custom_loss(true, pred):.4f}")
+
+model.compile(optimizer='adam', loss=custom_loss)
```
Comment (kalininei) [inline clock/train.py:124]: Мне бы хотелось видеть `learning_rate` в явном виде где-нибудь?

File: clock/train.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+model_checkpoint = callbacks.ModelCheckpoint(filepath='best_model.keras', save_best_only=True, monitor='val_loss')
+
+model.fit(
+    train_generator,
+    steps_per_epoch=len(train_data) // BATCH_SIZE,
+    validation_data=valid_generator ,
+    validation_steps=len(valid_data) // BATCH_SIZE,
+    epochs=EPOCHS,
+    callbacks=[tensorboard_callback, model_checkpoint]
+)
```
Comment (kalininei) [inline clock/train.py:136]: для формальности сохраните финальную версию.

File: clock/train.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+    delta_deg = y_pred_deg - y_true_deg
+
+    delta_deg = (delta_deg + 2160) % 4320 - 2160
+
+    return tf.reduce_mean(tf.square(delta_deg))
+
+train_data, valid_data = make_set(DATA_DIR)
+
+aug_sequence = lambda x: augmenter.get_augmenter(x)
```
[the remark this answers is not in the document]
Comment (kalininei) [reply clock/train.py:95]: А не, нужно `augmenter.get_augmenter(level)`
А эта функция вернёт функтор.

File: clock/augmenter.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
+import argparse
```
Comment (kalininei) [inline clock/augmenter.py:1]: не всё импорты нужны

File: clock/train.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+            batch_data.append(image)  
+            batch_labels.append(label)  
+
+        if batch_data:  
+            batch_data = np.array(batch_data, dtype=np.uint8) 
+            batch_labels = np.array(batch_labels) 
+
+            if aug_sequence is not None:
+                batch_data = aug_sequence(images=batch_data)  
+                augmenter.display_images(batch_data[:4], batch_labels[:4], 224, 224)
```
[the remark this answers is not in the document]
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    while True:
+        batch_data = []
+        batch_labels = []
+        
+        for _ in range(batch_size):
+            if len(data_copy) == 0: 
+                if need_shuffle:
+                    random.shuffle(data)  
+                data_copy = data.copy() 
+                continue
```
[the remark this answers is not in the document]
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    if need_shuffle:
+        random.shuffle(data_copy)
+
+    while True:
+        batch_data = []
+        batch_labels = []
+        
+        for _ in range(batch_size):
+            if len(data_copy) == 0: 
+                if need_shuffle:
```
[the remark this answers is not in the document]
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+            csv_file = os.path.join(clock_path, 'dataset_info.csv')
+
+            if os.path.isfile(csv_file):
+                df = pd.read_csv(csv_file)
+                for index, row in df.iterrows():
+                    image_name = row['imagepath']
+                    minutes = row['minutes']
+                    image_path = os.path.join(clock_path, image_name)
+                    all_data.append((image_path, minutes))
+    np.random.shuffle(all_data)
```
[the remark this answers is not in the document]
Comment (ol00lo) [reply clock/train.py:41]: ✔️

File: clock/train.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+from tensorflow.keras.applications import MobileNet
+from tensorflow.keras import layers, models, callbacks
+import augmenter
+
+IMG_HEIGHT, IMG_WIDTH = 224, 224
+BATCH_SIZE = 32
+EPOCHS = 10  
+DATA_DIR = 'c:\\Users\\mymri\\repos\\clock_sett\\'
+
+def make_set(DATA_DIR):
```
[the remark this answers is not in the document]
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+            batch_labels = np.array(batch_labels) 
+
+            if aug_sequence is not None:
+                batch_data = aug_sequence(images=batch_data)  
+                augmenter.display_images(batch_data[:4], batch_labels[:4], 224, 224)
+            
+            yield batch_data, batch_labels  
+
+def custom_loss(y_true, y_pred):
+    y_true_deg = tf.convert_to_tensor(y_true * 6, dtype=tf.float32)
```
[the remark this answers is not in the document]
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+    delta_deg = (delta_deg + 2160) % 4320 - 2160
+
+    return tf.reduce_mean(tf.square(delta_deg))
+
+train_data, valid_data = make_set(DATA_DIR)
+
+aug_sequence = lambda x: augmenter.get_augmenter(x)
+
+train_generator = data_generator(train_data, BATCH_SIZE, aug_sequence(0), True)
+valid_generator = data_generator(valid_data, BATCH_SIZE, lambda x: x, False)
```
[the remark this answers is not in the document]
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+
+model = models.Sequential([
+    base_model,
+    layers.GlobalAveragePooling2D(),
+    layers.Dense(128, activation='relu'),
+    layers.Dropout(0.5), 
+    layers.Dense(1, activation='linear')  
+])
+
+true = np.array([0.0, 720.0])
```
[the remark this answers is not in the document]
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+import time
+import random
+import tensorflow as tf
+from tensorflow.keras.applications import MobileNet
+from tensorflow.keras import layers, models, callbacks
+import augmenter
+
+IMG_HEIGHT, IMG_WIDTH = 224, 224
+BATCH_SIZE = 32
+EPOCHS = 10
```
[the remark this answers is not in the document]
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+print(f"36.0 = {custom_loss(true, pred):.4f}")
+
+model.compile(optimizer='adam', loss=custom_loss)
+log_dir = os.path.join("logs", "fit", f"model_run_{int(time.time())}")
+tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)
+model_checkpoint = callbacks.ModelCheckpoint(filepath='best_model.keras', save_best_only=True, monitor='val_loss')
+
+model.fit(
+    train_generator,
+    steps_per_epoch=len(train_data) // BATCH_SIZE,
```
[the remark this answers is not in the document]
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+pred = np.array([350.0])
+print(f"a lot = {custom_loss(true, pred)}")
+true = np.array([1.0])
+pred = np.array([2.0])
+print(f"36.0 = {custom_loss(true, pred):.4f}")
+
+model.compile(optimizer='adam', loss=custom_loss)
+log_dir = os.path.join("logs", "fit", f"model_run_{int(time.time())}")
+tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)
+model_checkpoint = callbacks.ModelCheckpoint(filepath='best_model.keras', save_best_only=True, monitor='val_loss')
```
[the remark this answers is not in the document]
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py (hunk 7 above)
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py (hunk 8 above)
Comment (ol00lo) [reply clock/train.py]: ✔️

File: clock/train.py (hunk 9 above)
Comment (ol00lo) [reply clock/train.py:129]: ✔️

File: clock/augmenter.py (hunk 11 above)
Comment (ol00lo) [reply clock/augmenter.py]: ✔️
